src/core/ddsi/src/ddsi_handshake.c

```diff
@@ -367,6 +379,17 @@ VALIDATION_FAILED |                                         |
 static const dds_security_fsm_transition handshake_transistions [] =
 {   /* Start */
     { NULL,                                 EVENT_AUTO,                                 NULL,
+                                            &state_initial_delay                           },
+    /* initial delay: a short delay to give the remote node some time for matching the
+       BuiltinParticipantStatelessMessageWriter, so that it won't drop the auth_request
+       we're sending (that would result in a time-out and the handshake taking longer
+       than required). For the node that receives the auth_request, the transition for
+       the event EVENT_RECEIVED_MESSAGE_REQUEST is added, because that node may receive the
+       auth_request during this delay and can continue immediately (as the sender already
+       waited for this delay before sending the request) */
+    { &state_initial_delay,                 EVENT_TIMEOUT,                              NULL,
+                                            &state_validate_remote_identity                },
+    { &state_initial_delay,                 EVENT_RECEIVED_MESSAGE_REQUEST,             NULL,
                                             &state_validate_remote_identity                },
```

Comment:
When receiving a handshake request while waiting for the initial timeout to finish it is may be better to have an additional state which handles both the validate_remote_identity and the received handshake request message, because otherwise the handshake request will not be handled and we have to wait for a resent of this request.

Reply:
You're right, the current approach leads to some overhead. In 1c85b868d95c33df9d3acf89fdc0900a68898be2 I've added the additional state that does both `validate_remote_identity` and `begin_handshake_reply`. 

I think that, in case of a transition from `state_initial_delay` to the newly introduced state `state_validate_remote_and_begin_reply` (which means the node received an `auth_request`) the result from `validate_remote_identity` should always be `PENDING_HANDSHAKE_MESSAGE` (or `FAILED` in case something went wrong). So I added a check on this return value and the handshake fails if this is not the case.